File: src/main/javascript/DoubleLinkedList.js
class _Node {
    _value
    _prev
    _next

    constructor(value) {
        this._value = value
    }
}

export class DoubleLinkedList {
    __head = new _Node(null)
    __tail = new _Node(null)
    __size = 0

    constructor() {
        this.__head._next = this.__tail
        this.__tail._prev = this.__head
    }

    getSize() {
        return this.__size
    }

    append(value) {
        const newNode = new _Node(value)
        newNode._prev = this.__tail._prev
        newNode._next = this.__tail
        this.__tail._prev._next = newNode
        this.__tail._prev = newNode

        this.__size++
    }

    appendLeft(value) {
        const newNode = new _Node(value)
        newNode._prev = this.__head
        newNode._next = this.__head._next
        this.__head._next._prev = newNode
        this.__head._next = newNode

        this.__size++
    }

    insert(index, value) {
        if (index === 0) {
            this.appendLeft(value)
        } else if (index === this.__size) {
            this.append(value)
        } else {
            const nextNode = this.__getNodeAt(index)
            const node = new _Node(value)

            nextNode._prev._next = node
            node._prev = nextNode._prev
            node._next = nextNode
            nextNode._prev = node

            this.__size++
        }
    }

    get(index) {
        return this.__getNodeAt(index)._value
    }

    removeAt(index) {
        const node = this.__getNodeAt(index)
        node._prev._next = node._next
        node._next._prev = node._prev

        this.__size--

        return node._value
    }

    __getNodeAt(index) {
        this.__checkIndex(index)

        let node
        if (index < this.__size / 2) {
            node = this.__head
            for (let i = 0; i < index + 1; i++) {
                node = node._next
            }
        } else {
            node = this.__tail
            for (let i = 0; i < this.__size - index; i++) {
                node = node._prev
            }
        }

        return node
    }

    __checkIndex(index) {
        if (index < 0 || this.__size <= index) {
            throw new RangeError()
        }
    }
}

File: src/main/python/DoubleLinkedList.py
from typing import Optional


class _Node:
    def __init__(self, value):
        self.value = value
        self.prev: Optional[_Node] = None
        self.next: Optional[_Node] = None


class SingleLinkedList:
    __head = _Node(value=None)
    __tail = _Node(value=None)
    __size = 0

    def __init__(self):
        self.__head.next = self.__tail
        self.__tail.prev = self.__head

    def get_size(self):
        return self.__size

    def append(self, value):
        new_node = _Node(value=value)
        new_node.prev = self.__tail.prev
        new_node.next = self.__tail
        self.__tail.prev.next = new_node
        self.__tail.prev = new_node

        self.__size += 1

    def append_left(self, value):
        new_node = _Node(value=value)
        new_node.prev = self.__head
        new_node.next = self.__head.next
        self.__head.next.prev = new_node
        self.__head.next = new_node

        self.__size += 1

    def insert(self, index, value):
        if index == 0:
            self.append_left(value=value)
        elif index == self.__size:
            self.append(value=value)
        else:
            next_node = self.__get_node_at(index=index)
            node = _Node(value=value)

            next_node.prev.next = node
            node.prev = next_node.prev
            node.next = next_node
            next_node.prev = node

            self.__size += 1

    def get(self, index):
        return self.__get_node_at(index=index).value

    def remove_at(self, index):
        node = self.__get_node_at(index=index)
        node.prev.next = node.next
        node.next.prev = node.prev

        self.__size -= 1

        return node.value

    def __get_node_at(self, index) -> _Node:
        self.__check_index(index=index)

        if index < self.__size / 2:
            node = self.__head
            for _ in range(index + 1):
                node = node.next
        else:
            node = self.__tail
            for _ in range(self.__size - index):
                node = node.prev

        return node

    def __check_index(self, index):
        if not (0 <= index < self.__size):
            raise IndexError
